src/cpp/Cdr.cpp

```diff
@@ -2697,67 +2216,1087 @@ Cdr& Cdr::deserializeBoolSequence(
+    current_state.next_member_id_ = member_id;
+    next_member_id_ = MEMBER_ID_INVALID;
 
-    numElements = seqLength;
     return *this;
 }
+
+Cdr& Cdr::xcdr1_end_serialize_opt_member(
+        const Cdr::state& current_state)
+{
+    assert(MEMBER_ID_INVALID != current_state.next_member_id_);
+    assert(EncodingAlgorithmFlag::PLAIN_CDR == current_encoding_ ||
+            EncodingAlgorithmFlag::PL_CDR == current_encoding_);
+
+    if (0 < current_state.member_size_)
+    {
+        const size_t member_serialized_size = offset_ - origin_;
+        set_state(current_state);
+        if (member_serialized_size > std::numeric_limits<uint16_t>::max())
+        {
+            switch (current_state.header_serialized_)
+            {
+                case XCdrHeaderSelection::SHORT_HEADER:
+                    if (AUTO_WITH_SHORT_HEADER_BY_DEFAULT == current_state.header_selection_)
+                    {
+                        xcdr1_change_to_long_member_header(current_state.next_member_id_, member_serialized_size);
+                    }
+                    else
+                    {
+                        throw BadParamException(
+                                  "Cannot encode XCDRv1 ShortMemberHeader when serialized member size is greater than 0xFFFF");
+                    }
+                    break;
+                case XCdrHeaderSelection::LONG_HEADER:
+                    xcdr1_end_long_member_header(current_state.next_member_id_, member_serialized_size);
+                    break;
+                default:
+                    assert(false); // header_serialized_ must have only SHORT_HEADER or LONG_HEADER
+            }
+        }
+        else
+        {
+            switch (current_state.header_serialized_)
+            {
+                case XCdrHeaderSelection::SHORT_HEADER:
+                    xcdr1_end_short_member_header(current_state.next_member_id_, member_serialized_size);
+                    break;
+                case XCdrHeaderSelection::LONG_HEADER:
+                    if (LONG_HEADER == current_state.header_selection_ ||
+                            0x3F00 < current_state.next_member_id_.id)
+                    {
+                        xcdr1_end_long_member_header(current_state.next_member_id_, member_serialized_size);
+                    }
+                    else if (AUTO_WITH_LONG_HEADER_BY_DEFAULT == current_state.header_selection_)
+                    {
+                        xcdr1_change_to_short_member_header(current_state.next_member_id_, member_serialized_size);
+                    }
+                    break;
+                default:
+                    assert(false); // header_serialized_ must have only SHORT_HEADER or LONG_HEADER
+            }
+        }
+        jump(member_serialized_size);
+    }
+    else if (EncodingAlgorithmFlag::PL_CDR != encoding_flag_)
+    {
+        // Reset state to POP(origin=0) because before serializing member the algorithm did PUSH(origin=0).
```

Comment:
How did you determine you needed this? I don't see any usage of `POP` in [the spec](https://www.omg.org/spec/DDS-XTypes/1.3/PDF)

Also are you taking the PUSH(origin=0) to mean that alignment now happens relative to the current offset? I was confused on this point when reading the spec but took it to mean origin should be set to offset and was wondering if that was correct.

Reply:
You're right. I fix it in 688010d. Also added a regression test for this case. Thanks